nimbus/core/block_import.nim

```
@@ -14,47 +14,55 @@ import
-    errorCount = 0
-    blk: array[1, EthBlock]
+    blk: EthBlock
 
   # even though the new imported blocks have block number
   # smaller than head, we keep importing it.
   # it maybe a side chain.
-  # TODO the above is no longer true with a single-state database - to deal with
-  #      that scenario the code needs to be rewritten to not persist the blocks
-  #      to the state database until all have been processed
   while rlp.hasData:
-    blk[0] = try:
+    blk = try:
       rlp.read(EthBlock)
     except RlpError as e:
       # terminate if there was a decoding error
-      error "rlp error",
-        fileName = importFile,
-        msg = e.msg,
-        exception = e.name
-      return false
-
-    chain.persistBlocks(blk).isOkOr():
-      # register one more error and continue
-      error "import error",
-        fileName = importFile,
-        error
-      errorCount.inc
-
-  return errorCount == 0
-
-proc importRlpBlock*(importFile: string; com: CommonRef): bool =
-  let res = io2.readAllBytes(importFile)
-  if res.isErr:
-    error "failed to import",
-      fileName = importFile
-    return false
-
-  importRlpBlock(res.get, com, importFile)
+      return err($e.name & ": " & e.msg)
+
+    ? chain.importBlock(blk)
+
+  if finalize:
+    ? chain.forkChoice(chain.latestHash, chain.latestHash)
+
+  ok()
+
+proc importRlpBlocks(importFile: string,
+                     chain: ForkedChainRef,
+                     finalize: bool): Result[void, string] =
+  let bytes = io2.readAllBytes(importFile).valueOr:
+    return err($error)
+  importRlpBlocks(bytes, chain, finalize)
+
+proc importRlpBlocks*(conf: NimbusConf, com: CommonRef) =
+  var head: BlockHeader
+  if not com.db.getCanonicalHead(head):
+    error "cannot get canonical head from db"
+    quit(QuitFailure)
+
+  let chain = newForkedChain(com, head, baseDistance = 0)
+
+  # success or not, we quit after importing blocks
+  for i, blocksFile in conf.blocksFile:
+    if isFile(string blocksFile):
```

Potential TOCTOU issue, also, doesn't check for permissions and some other issues? I'm not sure why not let the `io2.readAllBytes`-based error checking handle this, since this up-front check is both insufficient on its own and doesn't seem to provide particularly useful UX feedback.

Indeed, it silently skips `blockFiles` which aren't files or don't exist, but that seems to be debatably useful. If someone specifies a file to import from and that file does not exist, that seems like an unexpected condition.

---

https://github.com/status-im/nimbus-eth1/pull/2441/commits/e48855f437f60bd7767076ee83236207cdb4219c